sdk/src/client/api/mana.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    client::Client,
+    types::block::{
+        protocol::{lower_bits, multiplication_and_shift, upper_bits},
+        slot::SlotIndex,
+    },
+};
+
+impl Client {
+    /// Calculates the potential mana that is generated by holding `amount` tokens from `slot_index_created` to
+    /// `slot_index_target` and applies the decay to the result
+    pub async fn potential_mana(
+        &self,
+        amount: u64,
+        slot_index_created: SlotIndex,
+        slot_index_target: SlotIndex,
+    ) -> crate::client::Result<u64> {
+        if slot_index_created >= slot_index_target {
+            return Ok(0);
+        }
+        let protocol_parameters_map = self.get_info().await?.node_info.protocol_parameters;
+        let params: &crate::types::block::protocol::ProtocolParameters = &protocol_parameters_map.latest().parameters;
```

Comment:
Hmm, so then what if start and target epoch are using different protocol params? The whole calculation is potentially wrong in that case. 😱

Reply:
Yeah I guess it needs to be adjusted for that

Reply:
Maybe just have a function around this one, that loops over the protocol parameters with epoch index from when the output was created until this point and for each of them call this `potential_mana()`? Probably also worth to check what they did in go, for the node they need to calculate it also

Reply:
They simply don't handle the possibility of a protocol parameter change at all. This is their fn to convert from slot to epoch, which does not take into account a variable `slotsPerEpochExponent`.

```go
// EpochFromSlot calculates the EpochIndex from the given slot.
func (t *TimeProvider) EpochFromSlot(slot SlotIndex) EpochIndex {
	return EpochIndex(slot >> t.slotsPerEpochExponent)
}
```

So I'm really not sure what to do about this.

Reply:
Ok apparently it's something we shouldn't care about now, because it's not planned to have new protocol parameters without bigger changes, which we would need to implement before anyways